include/zephyr/sys/util.h

```diff
@@ -120,8 +120,70 @@ extern "C" {
  * @param array an array
  * @return 1 if @p ptr is part of @p array, 0 otherwise
  */
-#define PART_OF_ARRAY(array, ptr) \
-	((ptr) && ((ptr) >= &array[0] && (ptr) < &array[ARRAY_SIZE(array)]))
+#define PART_OF_ARRAY(array, ptr)                                                                  \
+	((ptr) && (uint8_t *)&(array)[0] <= (uint8_t *)(ptr) &&                                    \
+	 (uint8_t *)(ptr) < (uint8_t *)&(array)[ARRAY_SIZE(array)])
```

Comment:
General question: Why are we casting the pointer to a `uint8_t` in these macros btw? Wouldn't the pointer arithmetic work on any pointer value?

Reply:
For pointer arithmetic we should use `uintptr_t` through `POINTER_TO_UINT()`.

Reply:
That doesn't my answer as to why the cast to `uint8_t *` is useful or necessary :) 


Wouldn't `uintptr_t` also only be useful if we needed to store the result in a specific type? e.g. in `((ptr) - (array))` it doesn't seem necessary to cast the pointer to a `uintptr_t`

Reply:
> To avoid compiler warnings (comparing different pointer types).

Isn't different pointer types inherently an issue in `IS_ARRAY_ELEMENT`? Should we in that macro not then actively avoid casting the pointer type, as we do not wish to compare pointers of different types without a compiler warning?

Reply:
> > To avoid compiler warnings (comparing different pointer types).
> 
> Isn't different pointer types inherently an issue in `IS_ARRAY_ELEMENT`? Should we in that macro not then actively avoid casting the pointer type, as we do not wish to compare pointers of different types without a compiler warning?

Casting one pointer type to another is fully legal in C and since this is an address-based comparison, casting (to `uintptr_t` for integer arithmetic) is perfectly justified.

Reply:
I don't disagree with that, but your argument
> To avoid compiler warnings (comparing different pointer types).
Is not very good since we, in `IS_ARRAY_ELEMENT` actually want to compare pointers of the same type. 

Either way it works :)